ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterSortPredicates.java

```diff
@@ -0,0 +1,268 @@
+          if (!call.getOperands().equals(newAndOperands)) {
+            modified = true;
+            return call.clone(call.getType(), newAndOperands);
+          }
+          break;
+        case OR:
+          List<RexNode> newOrOperands = call.getOperands()
+              .stream()
+              .map(pred -> new Pair<>(pred, rankingOr(pred)))
+              .sorted(Comparator.comparing(Pair::getValue, Comparator.nullsLast(Double::compare)))
+              .map(Pair::getKey)
+              .collect(Collectors.toList());
+          if (!call.getOperands().equals(newOrOperands)) {
+            modified = true;
+            return call.clone(call.getType(), newOrOperands);
+          }
+          break;
+      }
+      return call;
+    }
+
+    private Double rankingAnd(RexNode e) {
+      Double selectivity = selectivityEstimator.estimateSelectivity(e);
+      if (selectivity == null) {
+        return null;
+      }
+      Double costPerTuple = costPerTuple(e);
+      if (costPerTuple == null) {
+        return null;
+      }
+      return (selectivity - 1d) / costPerTuple;
+    }
+
+    private Double rankingOr(RexNode e) {
+      Double selectivity = selectivityEstimator.estimateSelectivity(e);
+      if (selectivity == null) {
+        return null;
+      }
+      Double costPerTuple = costPerTuple(e);
+      if (costPerTuple == null) {
+        return null;
+      }
+      return -selectivity / costPerTuple;
+    }
+
+    private Double costPerTuple(RexNode e) {
+      return e.accept(new RexFunctionCost());
+    }
+
+  }
+
+  /**
+   * The cost of a call expression e is computed as:
+   * cost(e) = functionCost + sum_1..n(byteSize(o_i) + cost(o_i))
+   * with the call having operands i in 1..n.
+   */
+  private static class RexFunctionCost extends RexVisitorImpl<Double> {
+
+    private RexFunctionCost() {
+      super(true);
+    }
+
+    @Override
+    public Double visitCall(RexCall call) {
+      if (!deep) {
+        return null;
+      }
+
+      Double cost = 0.d;
+      for (RexNode operand : call.operands) {
+        Double operandCost = operand.accept(this);
+        if (operandCost == null) {
+          return null;
+        }
+        cost += operandCost;
+        Double size = HiveRelMdSize.averageTypeSize(operand.getType());
+        if (size == null) {
+          return null;
+        }
+        cost += size;
+      }
+
+      return cost + functionCost(call);
+    }
+
+    private static Double functionCost(RexCall call) {
+      switch (call.getKind()) {
+        case EQUALS:
+        case NOT_EQUALS:
+        case LESS_THAN:
+        case GREATER_THAN:
+        case LESS_THAN_OR_EQUAL:
+        case GREATER_THAN_OR_EQUAL:
+        case IS_NOT_NULL:
+        case IS_NULL:
+        case IS_TRUE:
+        case IS_NOT_TRUE:
+        case IS_FALSE:
+        case IS_NOT_FALSE:
+          return 1d;
+
+        case BETWEEN:
+          return 3d;
+
+        case IN:
+          return 2d * (call.getOperands().size() - 1);
```

Comment:
Can you add comment to explain what is the rationale for choosing these "arbitrary constant" for function costs.

Reply:
Done!